// 3id-did-provider/lib/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toStableObject = exports.decodeBase64 = exports.encodeBase64 = exports.u8aToHex = exports.hexToU8A = exports.parseJWEKids = exports.decodeKey = exports.encodeKey = void 0;
const fast_json_stable_stringify_1 = __importDefault(require("fast-json-stable-stringify"));
const u8a = __importStar(require("uint8arrays"));
const B16 = 'base16';
const B64 = 'base64pad';
const multicodecPubkeyTable = {
    secp256k1: 0xe7,
    x25519: 0xec,
    ed25519: 0xed,
};
function encodeKey(key, keyType) {
    const bytes = new Uint8Array(key.length + 2);
    if (!multicodecPubkeyTable[keyType]) {
        throw new Error(`Key type "${keyType}" not supported.`);
    }
    bytes[0] = multicodecPubkeyTable[keyType];
    bytes[1] = 0x01;
    bytes.set(key, 2);
    return `z${u8a.toString(bytes, 'base58btc')}`;
}
exports.encodeKey = encodeKey;
function decodeKey(key) {
    const bytes = u8a.fromString(key.slice(1), 'base58btc');
    const supportedKey = bytes[1] === 0x01 &&
        (multicodecPubkeyTable['secp256k1'] === bytes[0] ||
            multicodecPubkeyTable['x25519'] === bytes[0] ||
            multicodecPubkeyTable['ed25519'] === bytes[0]);
    if (!supportedKey)
        throw new Error(`Key type ${bytes[0]} not supported`);
    return bytes.slice(2);
}
exports.decodeKey = decodeKey;
function parseJWEKids(jwe) {
    var _a;
    return (((_a = jwe.recipients) === null || _a === void 0 ? void 0 : _a.reduce((kids, recipient) => {
        var _a;
        if ((_a = recipient.header) === null || _a === void 0 ? void 0 : _a.kid)
            kids.push(recipient.header.kid.split('#')[1]);
        return kids;
    }, [])) || []);
}
exports.parseJWEKids = parseJWEKids;
function hexToU8A(s) {
    return u8a.fromString(s, B16);
}
exports.hexToU8A = hexToU8A;
function u8aToHex(b) {
    return u8a.toString(b, B16);
}
exports.u8aToHex = u8aToHex;
function encodeBase64(b) {
    return u8a.toString(b, B64);
}
exports.encodeBase64 = encodeBase64;
function decodeBase64(s) {
    return u8a.fromString(s, B64);
}
exports.decodeBase64 = decodeBase64;
function toStableObject(obj) {
    return JSON.parse(fast_json_stable_stringify_1.default(obj));
}
exports.toStableObject = toStableObject;
